docs/WitnessMatchingManifesto.md

```diff
@@ -0,0 +1,367 @@
+# Protocol Witness Matching Manifesto
+
+Author: [Suyash Srijan](https://www.github.com/theblixguy) 
+
+### Jargon
+
+Before we dive into the document, let's clarify some jargon:
+
+- A _protocol requirement_ (or just _requirement_) is a declaration inside a protocol that all conforming types must satisfy.
+- A _protocol witness_ (or just _witness_) is a value or a type that satisfies a protocol requirement.
+
+For example:
+
+```swift
+protocol P {
+  func foo() // This is a protocol requirement
+}
+
+struct S: P {
+  func foo() { // This is a protocol witness
+    ...
+  }
+}
+```
+
+## Introduction
+
+In Swift (at the time of this writing), protocol witnesses have to pretty much match exactly with a protocol requirement. For example:
+
+```swift
+protocol P {
+  func foo()
+}
+
+struct S: P {
+  func foo() { ... }
+}
+```
+
+```swift
+protocol P {
+  func getView() -> UIView
+}
+
+struct S: P {
+  func getView() -> UIView { ... }
+}
+```
+
+A few specific differences are allowed:
+
+1. A non-throwing function can witness a throwing function requirement:
+
+```swift
+protocol P {
+  func foo() throws
+}
+
+struct S: P {
+  // Even though 'foo' here is not marked 'throws', it is
+  // still able to satisfy the protocol requirement. The
+  // caller has to handle the possibility that it might 
+  // throw, but the witness is ofcourse free to throw or 
+  // not to, depending on whether it makes sense in the
+  // implementation.
+  func foo() { ... }
+}
+```
+
+2. A non-failable initializer can witness a failable initializer requirement:
```

Comment:
infallible and fallible instead of non-failable and failable?

Reply:
[TSPL](https://docs.swift.org/swift-book/LanguageGuide/Initialization.html) uses the word "failable" and "nonfailable".

Reply:
“Fallible” and “failable” aren’t synonymous.

Reply:
Huh, TIL. Thanks!

Reply:
@xwu what is the difference? If I look at [Merriam-Webster](https://www.merriam-webster.com/dictionary/fallible):

> 1 : liable to be erroneous a fallible generalization
> 2 : capable of making a mistake

There is no definition for failable but I suppose all it means is "something that can fail" which seems to coincide with 2.

Reply:
The diagnostics and TSPL both use "non-failable" / "failable" (and personally I have never come across anyone saying "fallible"/"infallible" initializer), so even if there are similarities, I think it's best to stick to convention here.

Reply:
Yeah I'm not saying that we should diverge from the convention (which is why I marked the conversation as "resolved") but I'm curious what the difference is, since @xwu pointed out that there is some difference in meaning. [Sorry to hijack the post review 😅.]